### PR TITLE
SendDailyReport task

### PR DESCRIPTION
**Pull request by apereverzin**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-1026

### Change description ###
SendDailyReportTask implementing scheduled job to send daily reports.
Logical expression in @ConditionalOnExpression annotation enables SendDailyReportTask if it is enabled AND EmailSender bean is present i.e. spring.mail.host property is properly configured

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/report/ReportSender.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import java.io.File;
+import java.time.LocalDate;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+@Component
+@ConditionalOnProperty(prefix = "spring.mail", name = "host")
+public class ReportSender {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/report/ReportSender.java]: should this follow other scheduled tasks naming convention? (and package)

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/report/ReportSender.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import uk.gov.hmcts.reform.blobrouter.services.email.EmailSender;
+
+import java.io.File;
+import java.time.LocalDate;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+@Component
+@ConditionalOnProperty(prefix = "spring.mail", name = "host")
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/report/ReportSender.java]: how about a boolean flag (I think there's one for all other jobs)

File: src/main/resources/application.yaml [hunk 3]
```diff
@@ -70,6 +70,9 @@ service:
       target-storage-account: bulkscan
       target-container: sscs
 
+reports:
+  recipients: ${REPORTS_RECIPIENTS:}
```
Comment (rkondratowicz) [inline src/main/resources/application.yaml]: is `:` needed?

File: src/main/resources/application.yaml [hunk 4]
```diff
@@ -94,7 +97,8 @@ scheduling:
       cron: 0 0 6 * * *
     scan:
       delay: ${TASK_SCAN_DELAY} # In milliseconds
-
+    send-daily-report:
+      cron: ${SEND_DAILY_REPORT_CRON:0 0 18 ? * MON-FRI}
```
Comment (rkondratowicz) [inline src/main/resources/application.yaml]: missing `enabled:` I think

File: infrastructure/main.tf [hunk 5]
```diff
@@ -145,4 +145,19 @@ resource "azurerm_key_vault_secret" "db_flyway_password" {
...
+
+data "azurerm_key_vault_secret" "smtp_password" {
+  key_vault_id = "${data.azurerm_key_vault.reform_scan_key_vault.id}"
+  name         = "reports-email-password"
+}
+
+data "azurerm_key_vault_secret" "reports_recipients" {
+  key_vault_id = "${data.azurerm_key_vault.reform_scan_key_vault.id}"
+  name         = "reports-recipients"
+}
```
Comment (doncem) [inline infrastructure/main.tf]: i don't think we need this section at all

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        if (this.recipients.length == 0) {
+            logger.warn("No recipients configured for reports");
+        }
+    }
+    // endregion
+
+    @Scheduled(cron = "${scheduling.task.send-daily-report.cron}")
+    @SchedulerLock(name = "report-sender")
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java]: it's actually `daily summary report` and not just some unknown `report sender`. sounds good?

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import java.io.File;
+import java.time.LocalDate;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+@Component
+@ConditionalOnExpression("!'${spring.mail.host}'.equals(null) "
+    + "&& !'${spring.mail.host}'.equals('false') "
+    + "&& ${scheduling.task.send-daily-report.enabled:true}")
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java]: let's just check the `enabled` flag (like in other scheduled jobs) if possible.

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java]: @rkondratowicz Done

File: src/main/resources/application.yaml (hunk 4 above)
Comment (apereverzin) [reply src/main/resources/application.yaml]: @rkondratowicz Done

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/email/EmailSender.java [hunk 8]
```diff
@@ -12,7 +12,7 @@
 import javax.mail.internet.MimeMessage;
 
 @Component
-@ConditionalOnProperty(prefix = "spring.mail", name = "host")
+@ConditionalOnProperty("spring.mail.host")
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/email/EmailSender.java]: i think this was unnecessary change

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+import uk.gov.hmcts.reform.blobrouter.services.report.ReportService;
+
+import java.io.File;
+import java.time.LocalDate;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+@Component
+@ConditionalOnProperty("scheduling.task.send-daily-report.enabled")
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java]: i think we usually have this defined in slightly different format is it not?

File: infrastructure/main.tf [hunk 10]
```diff
@@ -125,4 +125,25 @@ resource "azurerm_key_vault_secret" "POSTGRES_DATABASE" {
   key_vault_id = "${data.azurerm_key_vault.reform_scan_key_vault.id}"
   value        = "${module.reform-blob-router-db.postgresql_database}"
 }
+
+resource "azurerm_key_vault_secret" "db_flyway_password" {
+  name         = "db-flyway-password"
+  value        = "${module.reform-blob-router-db.postgresql_password}"
+  key_vault_id = "${data.azurerm_key_vault.reform_scan_key_vault.id}"
+}
```
Comment (rkondratowicz) [inline infrastructure/main.tf]: unrelated to reports

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java (hunk 6 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java]: let's use `@SchedulerLock(name = TASK_NAME)` as all other scheduled tasks do (and also add logs on job starting in finishing)

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+public class SendDailyReportTask {
+    private static final Logger logger = LoggerFactory.getLogger(SendDailyReportTask.class);
+
+    static final String EMAIL_SUBJECT = "Reform Scan daily report";
+    static final String EMAIL_BODY = "This is an auto generated email. Do not respond to it.";
+    static final String ATTACHMENT_PREFIX = "reform_scan_envelopes_";
+    static final String ATTACHMENT_SUFFIX = ".csv";
+
+    private static final String TASK_NAME = "report-sender";
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java]: for all other tasks this matches class name

File: infrastructure/main.tf [hunk 12]
```diff
@@ -126,3 +126,26 @@ resource "azurerm_key_vault_secret" "POSTGRES_DATABASE" {
...
+
+# endregion
+
+# region reports  secrets
+
+data "azurerm_key_vault_secret" "reports_recipients" {
+  key_vault_id = "${data.azurerm_key_vault.reform_scan_key_vault.id}"
+  name         = "reports-recipients"
+}
```
Comment (lgonczar) [inline infrastructure/main.tf]: It looks like passing these secrets to the application is missing.

Before merging this PR, please make sure these secrets are set in vaults for aat, prod, demo and perftest.

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/features/SchedulerConfigTest.java [hunk 13]
```diff
@@ -37,7 +39,7 @@ public void should_integrate_with_shedlock() throws Exception {
         ArgumentCaptor<LockConfiguration> configCaptor = ArgumentCaptor.forClass(LockConfiguration.class);
 
         // wait for asynchronous run of the scheduled task in background
-        Thread.sleep(2000);
+        Thread.sleep(5000);
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/features/SchedulerConfigTest.java]: Was there a need for this increase? 5 seconds is quite a lot of waiting.

Comment (apereverzin) [reply src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/features/SchedulerConfigTest.java]: I have got one failure locally (not in Jenkins), this increase made it stable. Multithreaded tests are generally not very stable

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.List;
+import java.util.Map;
+
+@Component
+@ConditionalOnProperty("scheduling.task.send-daily-report.enabled")
+public class SendDailyReportTask {
+    private static final Logger logger = LoggerFactory.getLogger(SendDailyReportTask.class);
+
+    static final String EMAIL_SUBJECT = "Reform Scan daily report";
+    static final String EMAIL_BODY = "This is an auto generated email. Do not respond to it.";
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java]: I can see that this message is taken from bulk-scan-processor, but we should go for `autogenerated` or `auto-generated`, as it's incorrect as two words.

File: src/main/resources/application.yaml (hunk 3 above)
Comment (lgonczar) [reply src/main/resources/application.yaml]: We should only use default values where reasonable default values exist. I guess empty recipients field is always invalid in an email.

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/features/SchedulerConfigTest.java (hunk 13 above)
Comment (lgonczar) [reply src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/features/SchedulerConfigTest.java]: If we got to 5 seconds of waiting, maybe we should move to some polling-based approach with timeout, i.e. wait only as long as it's necessary?

File: src/main/resources/application.yaml (hunk 3 above)
Comment (apereverzin) [reply src/main/resources/application.yaml]: Will do

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java (hunk 14 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java]: Will do

File: src/main/resources/application.yaml (hunk 3 above)
Comment (lgonczar) [reply src/main/resources/application.yaml]: Ok, it might make sense to have an empty recipient list by default. It will only make sense, though, when host is not configured (as we make the job's presence dependent on it).

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+        this.from = from;
+
+        if (recipients == null) {
+            this.recipients = new String[0];
+        } else {
+            this.recipients = Arrays.copyOf(recipients, recipients.length);
+        }
+
+        if (this.recipients.length == 0) {
+            logger.warn("No recipients configured for reports");
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java]: I think it's an error situation and it would be much safer to throw an exception at this point to prevent the application from starting. Otherwise, we'll never send a report to anyone and won't even notice.

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java]: Agree

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/features/SchedulerConfigTest.java (hunk 13 above)
Comment (apereverzin) [reply src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/features/SchedulerConfigTest.java]: Reverted back to 2000

File: infrastructure/main.tf (hunk 12 above)
Comment (doncem) [reply infrastructure/main.tf]: these resources will not be used as part of terraform setup - this file change is completely not needed. secrets need to be set manually to the vault and charts, i believe, already requesting them to be present (i recall there being a separate pr)

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java (hunk 11 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/SendDailyReportTask.java]: missed the update to the TASK_NAME value?

File: src/main/resources/bootstrap.yaml [hunk 16]
```diff
@@ -16,6 +16,8 @@ spring:
         crime-storage-connection-string: storage.crime.connection-string
         reports-email-username: SMTP_USERNAME
         reports-email-password: SMTP_PASSWORD
+        reports-recipients: REPORTS_RECIPIENTS
+
```
Comment (doncem) [inline src/main/resources/bootstrap.yaml]: not sure it is a good idea to have empty line in bootstrap config here :thinking:

File: src/main/resources/application.yaml [hunk 17]
```diff
@@ -94,7 +97,9 @@ scheduling:
...
```
Comment (doncem) [inline src/main/resources/application.yaml]: is default needed? i think we don't provide defaults anymore

**Comment by apereverzin** [comment]
https://github.com/hmcts/blob-router-service/pull/345
